cmd/bootstrap/cmd/seal.go

```diff
@@ -43,9 +44,24 @@ func constructRootResultAndSeal(
 		RandomSource:       getRandomSource(flagBootstrapRandomSeed),
 	}
 
+	qcsWithSignerIDs := make([]*flow.QuorumCertificateWithSignerIDs, 0, len(clusterQCs))
```

Comment:
## Why adding `QuorumCertificateWithSignerIDs` type?

The smart contract emits a EpochCommit Event, in which the [ClusterQC contains `VoterIDs`](https://github.com/onflow/flow-go/blob/master/model/convert/service_event.go#L358-L363), and we need to convert the cadence value into golang struct to be used in protocol. It's better to keep the data type of `ClusterQC` similar to the same struct in smart contract, therefore,[ I had to revert the change to `ClusterQC` in order to keep using `VoterIDs` field,](https://github.com/onflow/flow-go/pull/2101/files#diff-1efaf8fd528651c4c22e9586d64243dae87ddb4efb83e5c175f621153f18852aR132), which makes the conversion side-effect free.

However the QC now uses SignerIndices, which means `flow.ClusterQCVoteDatasFromQCs` can't convert directly QC into ClusterQC. 

I could change `flow.ClusterQCVoteDatasFromQCs` to let it take `(qc, signerIDs)` , but it looks weird that the two inputs seems have no relation, but in fact the signerIDs must be converted from qc.SignerIndices.

In order to make it a bit type-safe, I decided to define the `QuorumCertificateWithSignerIDs`. Now the caller of `ClusterQCVoteDatasFromQCs` is responsible to convert QC into `QuorumCertificateWithSignerIDs`, such that clusterQC can be converted from it. 

The drawback of this approach is that now we have two QC definition `QuorumCertificate` and `QuorumCertificateWithSignerIDs`. Also weird. 

Thoughts?

Reply:
👍 
I like this option, because we anyway package the bootstrapping information on its own. By allowing the bootstrapping information to be in its own custom format, we gain flexibility.